harper-core/src/linting/duplicate_punctuation.rs

```diff
@@ -0,0 +1,197 @@
+                            _ => break,
+                        }
+                    }
+                    if kinds.len() > 1 {
+                        let span = toks[start].span.expanded_to_include(toks[end].span.end - 1);
+                        lints.push(Lint {
+                            span,
+                            lint_kind: LintKind::Punctuation,
+                            suggestions: build_suggestions(&kinds),
+                            message: "Remove duplicate or mixed punctuation.".to_owned(),
+                            priority: 63,
+                        });
+                    }
+                    i = end + 1;
+                    continue;
+                }
+            }
+            i += 1;
+        }
+        lints
+    }
+
+    fn description(&self) -> &str {
+        "Flags duplicate or mixed punctuation sequences."
+    }
+}
+
+fn is_interest(p: Punctuation) -> bool {
+    matches!(
+        p,
+        Punctuation::Comma
+            | Punctuation::Semicolon
+            | Punctuation::Colon
+            | Punctuation::Bang
+            | Punctuation::Question
+            | Punctuation::ForwardSlash
+            | Punctuation::Ampersand
+    )
+}
+
+fn in_same_group(first: Punctuation, next: Punctuation) -> bool {
+    match first {
+        Punctuation::Bang | Punctuation::Question => {
+            matches!(next, Punctuation::Bang | Punctuation::Question)
+        }
+        _ => first == next,
+    }
+}
+
+fn build_suggestions(seq: &[Punctuation]) -> Vec<Suggestion> {
+    use Punctuation::*;
+    let mut opts = match seq[0] {
+        Bang | Question => vec![Bang, Question],
+        other => vec![other],
+    };
+    opts.dedup();
+    opts.into_iter()
+        .map(|p| Suggestion::ReplaceWith(vec![punct_char(p)]))
+        .collect()
+}
+
+fn punct_char(p: Punctuation) -> char {
+    use Punctuation::*;
+    match p {
+        Comma => ',',
+        Semicolon => ';',
+        Colon => ':',
+        Bang => '!',
+        Question => '?',
+        ForwardSlash => '/',
+        Ampersand => '&',
+        _ => unreachable!(),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::linting::tests::assert_lint_count;
+
+    use super::DuplicatePunctuation;
+
+    #[test]
+    fn flags_double_comma() {
+        assert_lint_count("Wait,, what happened?", DuplicatePunctuation, 1);
+    }
+
+    #[test]
+    fn flags_double_semicolon() {
+        assert_lint_count("He hesitated;; then spoke.", DuplicatePunctuation, 1);
+    }
+
+    #[test]
+    fn flags_double_colon() {
+        assert_lint_count("Choices:: A or B.", DuplicatePunctuation, 1);
+    }
+
+    #[test]
+    fn flags_double_bang() {
+        assert_lint_count("Stop!!", DuplicatePunctuation, 1);
+    }
+
+    #[test]
+    fn flags_double_question() {
+        assert_lint_count("Really??", DuplicatePunctuation, 1);
+    }
+
+    #[test]
+    fn flags_mixed_qbang_pair() {
+        assert_lint_count("What?!", DuplicatePunctuation, 1);
```

Comment:
Maybe interrobang character could be suggested.

https://en.wikipedia.org/wiki/Interrobang

Reply:
As with `??` this is common but informal so some will want to flag anything informal and some won't. It's not "duplicate" though. Maybe "punctuation sequence" or "punctuation clusters" would be a more accurate name?

Reply:
I feel like detecting this should be in the current rule

- ,,
- ;;
- ::
- //
- .,
- :,
- ;:
- ,.
- ?.
- !.
- and other combination

And these in another 

- ??
- ?!
- !!!

So the rule can be disabled.

Also what about `...` does the rule reports it? Don't you have a rule for using `…` instead ?

Reply:
What is the status about this?